### ISSUE TITLE
Ctor fails to translate

### ISSUE THREAD

**Issue by virtualritz** [anderslanglands/bbl-build-rs#4]
Checkout [this branch](https://github.com/phyrondev/openimageio/tree/io_proxy_bbl_issue) of my OIIO Rust wrapper.

The `IOProxy` `Ctor` does not want to build:
```cpp
BBL_MODULE(oiio) {
    bbl::Enum<OIIO::Filesystem::IOProxy::Mode>();

    bbl::Class<OIIO::Filesystem::IOProxy>()
        .ctor(bbl::Class<OIIO::Filesystem::IOProxy>::Ctor<OIIO::string_view, OIIO::Filesystem::IOProxy::Mode>("file_name", "mode"), "ctor");

}
```

The resp. part of `filesystem.h` looks like this:
```cpp
class OIIO_UTIL_API IOProxy {
public:
    enum Mode { Closed = 0, Read = 'r', Write = 'w' };
    IOProxy () {}
    IOProxy (string_view filename, Mode mode)
        : m_filename(filename), m_mode(mode) {}
        
    ...
```

The `Ctor` part errors out with:
```
error: failed to run custom build command for `openimageio v0.1.0 (/home/moritz/code/crates/bbl-oiio)`

Caused by:
  process didn't exit successfully: `/home/moritz/code/crates/bbl-oiio/target/debug/build/openimageio-6a08d69e3843cb7f/build-script-build` (exit status: 101)
  --- stdout
  cargo::rustc-link-arg=-lstdc++
  running "cmake" "-G" "Ninja" "-DCMAKE_BUILD_TYPE=Release" "-S bbl-oiio" "-B /home/moritz/code/crates/bbl-oiio/target/debug/build/openimageio-b13a4227d4d54671/out/build" "-DCMAKE_INSTALL_PREFIX=/home/moritz/code/crates/bbl-oiio/target/debug/build/openimageio-b13a4227d4d54671/out" "-DBBL_LANGUAGES=rust"
  -- Configuring done (0.0s)
  -- Generating done (0.0s)
  -- Build files have been written to: /home/moritz/code/crates/bbl-oiio/target/debug/build/openimageio-b13a4227d4d54671/out/build
  running "cmake" "--build" "/home/moritz/code/crates/bbl-oiio/target/debug/build/openimageio-b13a4227d4d54671/out/build" "--target" "install"
  [1/7] Building CXX object CMakeFiles/bind-dummy.dir/bind/filesystem.cpp.o
  FAILED: CMakeFiles/bind-dummy.dir/bind/filesystem.cpp.o 
  /usr/bin/c++  -isystem /usr/include/opencv4 -isystem /usr/include/Imath -O3 -DNDEBUG -MD -MT CMakeFiles/bind-dummy.dir/bind/filesystem.cpp.o -MF CMakeFiles/bind-dummy.dir/bind/filesystem.cpp.o.d -o CMakeFiles/bind-dummy.dir/bind/filesystem.cpp.o -c /home/moritz/code/crates/bbl-oiio/bbl-oiio/bind/filesystem.cpp
  In file included from /home/moritz/code/crates/bbl-oiio/bbl-oiio/bind/filesystem.cpp:1:
  /usr/local/include/babble: In instantiation of ‘bbl::Class<C> bbl::Class<C>::ctor(Ctor<Args ...>, const char*) [with T = OpenImageIO_v2_4::Filesystem::IOProxy; Args = {OpenImageIO_v2_4::basic_string_view<char, std::char_traits<char> >, OpenImageIO_v2_4::Filesystem::IOProxy::Mode}; C = OpenImageIO_v2_4::Filesystem::IOProxy]’:
  /home/moritz/code/crates/bbl-oiio/bbl-oiio/bind/filesystem.cpp:10:14:   required from here
  /usr/local/include/babble:107:58: error: static assertion failed: can not construct class with these arguments
    107 |         static_assert(std::is_constructible<T, Args...>::value,
        |                                                          ^~~~~
  /usr/local/include/babble:107:58: note: ‘std::integral_constant<bool, false>::value’ evaluates to false
  [2/7] Generating oiio-c.cpp
  FAILED: oiio-c.cpp /home/moritz/code/crates/bbl-oiio/target/debug/build/openimageio-b13a4227d4d54671/out/build/oiio-c.cpp 
  cd /home/moritz/code/crates/bbl-oiio/target/debug/build/openimageio-b13a4227d4d54671/out/build && /usr/local/bin/bbl-translate /home/moritz/code/crates/bbl-oiio/bbl-oiio/bind/bbl-misc.cpp /home/moritz/code/crates/bbl-oiio/bbl-oiio/bind/attrdelegate.cpp /home/moritz/code/crates/bbl-oiio/bbl-oiio/bind/deepdata.cpp /home/moritz/code/crates/bbl-oiio/bbl-oiio/bind/filesystem.cpp /home/moritz/code/crates/bbl-oiio/bbl-oiio/bind/imagebuf.cpp /home/moritz/code/crates/bbl-oiio/bbl-oiio/bind/imagebufalgo.cpp /home/moritz/code/crates/bbl-oiio/bbl-oiio/bind/imagecache.cpp /home/moritz/code/crates/bbl-oiio/bbl-oiio/bind/imageio.cpp /home/moritz/code/crates/bbl-oiio/bbl-oiio/bind/paramlist.cpp /home/moritz/code/crates/bbl-oiio/bbl-oiio/bind/typedesc.cpp /home/moritz/code/crates/bbl-oiio/bbl-oiio/bind/ustring.cpp -- -std=c++17 -fsyntax-only -fno-spell-checking -isystem/usr/include/c++/13 -isystem/usr/include/x86_64-linux-gnu/c++/13 -isystem/usr/include/c++/13/backward -isystem/usr/lib/gcc/x86_64-linux-gnu/13/include -isystem/usr/local/include -isystem/usr/include/x86_64-linux-gnu -isystem/usr/include -I/usr/local/lib/resource/include -I/usr/local/include -I/usr/local/include -I/home/moritz/code/crates/bbl-oiio/bbl-oiio -I/usr/include -I/usr/include -I/usr/include/opencv4 -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include/Imath NOMINMAX -DNOMINMAX -D_MT -Wno-macro-redefined -- oiio -o /home/moritz/code/crates/bbl-oiio/target/debug/build/openimageio-b13a4227d4d54671/out/build -l rust
  In file included from /home/moritz/code/crates/bbl-oiio/bbl-oiio/bind/filesystem.cpp:1:
  /usr/local/include/babble:107:23: error: static assertion failed due to requirement 'std::is_constructible<OpenImageIO_v2_4::Filesystem::IOProxy, OpenImageIO_v2_4::basic_string_view<char, std::char_traits<char>>, OpenImageIO_v2_4::Filesystem::IOProxy::Mode>::value': can not construct class with these arguments
    107 |         static_assert(std::is_constructible<T, Args...>::value,
        |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /home/moritz/code/crates/bbl-oiio/bbl-oiio/bind/filesystem.cpp:10:10: note: in instantiation of function template specialization 'bbl::Class<OpenImageIO_v2_4::Filesystem::IOProxy>::ctor<OpenImageIO_v2_4::Filesystem::IOProxy, OpenImageIO_v2_4::basic_string_view<char>, OpenImageIO_v2_4::Filesystem::IOProxy::Mode>' requested here
     10 |         .ctor(bbl::Class<OIIO::Filesystem::IOProxy>::Ctor<OIIO::string_view, OIIO::Filesystem::IOProxy::Mode>("file_name", "mode"), "ctor");
        |          ^
  [2024-08-21 15:37:46.223] [error] [api_cpp_context.cpp:45] /home/moritz/code/babble/bbl/src/cpp_context.cpp:3209 compilation generated 1 errors. Cannot continue
  [2024-08-21 15:37:46.224] [error] [bbl-translate.cpp:489] extraction failed after 17.681196896000003s
  ninja: build stopped: subcommand failed.
```

What am I missing here?

**Comment by virtualritz** [comment]
I also tried constructing the `IOProxy`-derived classes.
I.e.`IOFile`, `IOVecOutput` and `IOMemReader`. Even trying to get their resp. default constructor to translate results in the above error.

**Comment by virtualritz** [comment]
I just saw [**Subclassing c++ Types in-C**](https://github.com/anderslanglands/babble?tab=readme-ov-file#subclassing-c-types-in-c).

Is that the issue? Is there a workaround?
Should I switch back to `bindgen` for these types?

**Comment by anderslanglands** [comment]
The IOProxy ctor is erroring because IOProxy is abstract: you can't construct one at all.

How are you trying to bind IOFile etc? this works for me:
```
    bbl::Enum<OIIO::Filesystem::IOProxy::Mode>();
    bbl::ClassIncomplete<FILE>();
    bbl::Class<OIIO::Filesystem::IOFile>()
        .ctor(bbl::Class<OIIO::Filesystem::IOFile>::Ctor<FILE*, OIIO::Filesystem::IOProxy::Mode>("file", "mode"), "new")
    ;
```

**Comment by anderslanglands** [comment]
The other constructors both take either a string view or a wstring, so you'd need to bind those types in order to be able to bind the ctor, or wrap the constructors with a free function in `bblext` that takes a `char const*` and passes it to the actual constructor (I'd suggest doing this)

**Comment by anderslanglands** [comment]
I've just pushed an update that gives a better error message in your original case:
```
/home/anders/packages/babble/0.8.0/include/babble: In instantiation of ‘bbl::Class<C>::Ctor<Args>::Ctor(CtorArgs ...) [with CtorArgs = {const char*, const char*}; Args = {OpenImageIO_v2_3::string_view, OpenImageIO_v2_3::Filesystem::IOProxy::Mode}; C = OpenImageIO_v2_3::Filesystem::IOProxy]’:
/home/anders/code/bbl-oiio/bind/imageio.cpp:290:130:   required from here
/home/anders/packages/babble/0.8.0/include/babble:91:49: error: static assertion failed: cannot construct an abstract class
   91 |             static_assert(std::is_abstract_v<C> == false, "cannot construct an abstract class");
      |                           ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
```

**Comment by virtualritz** [comment]
> The IOProxy ctor is erroring because IOProxy is abstract: you can't construct one at all.

Yeah, I figured this when I wrapped `TextureSystem`, yesterday, as that is abstract too.

The error I saw persisted because I had the wrapper for the default constructor there and `IOFile` does not have this constructor. Go figure! :facepalm: 

Once I commented that out, everything worked.

Sorry about the noise and thanks heaps for the ergonomics improvement in `babble`. :grin:
